posts/using-swagger-to-build-effective-apis.md

```diff
@@ -0,0 +1,382 @@
+{
+  "id" : 0,
+  "sets" : [ {
+    "id" : 0,
+    "exercise" : {
+      "id" : 0,
+      "name" : "string",
+      "description" : "string",
+      "primaryMuscleGroupId" : {
+        "id" : 0,
+        "name" : "string"
+      },
+      "movementModifier" : "isolation"
+    },
+    "repetitions" : [ {
+      "id" : 0,
+      "weight" : 0,
+      "system" : true,
+      "repetitionModified" : "tempo"
+    } ],
+    "name" : "string"
+  } ],
+  "avgRestTime" : 0
+}
+```
+We defined the payload as an exercise containing just default values for each field, and passed it in JSON to the curl statement under the `-d` flag.  As you can see, our API returned a sample exercise object that deviated from our payload.  It is difficult to see at first, but if you look closely you will see we passed an exercise with an id of "1" to the curl statment, and received an exercise with an id of "0" in the response body.  This is a short-coming inherent to mocking.  In our API specification, we may have intended our API to return the exact response we sent, but there is no way to emulate this behavior in a mock.  This is because REST APIs are stateful.  There is no way for us to emulate this intended behavior without designing business logic.  Situations like these are important to pay close attention to when mocking an API, as they may reveal flaws in your assumption about how your API will work.  
+
+# Summary
+Swagger is a great toolset for anyone looking to develop an API.  For those of us looking to write APIs, you should always adhere to a common standard which will make your API usable.  That is why you should use the OAS 3.0.n specification for your API.  APIs that do not adhere to this design have the potential to be difficult to use and maintain.  Any tool that is difficult to maintain will not be used.  Inversely, Swagger is a tool that is very easy to use, which is why I have developed this demo API using Swagger tools in the first place.  Some of the additional benefits of Swagger tools include their portability.  Swagger Hub will let you generate templated client and server code in nearly any language you want.  This allows you to rapidly prototype APIs for the future.  AWS API Gateway will generate help you to test your API, or even deploy your API in a serverless context using AWS Lambda to handle endpoint calls.  Maven has a swagger plug-in which takes a definition YAML file and generates client models on the fly so you can rapidly prototype an application for your API.  If the API changes, you just update the definition file and your API code will change with it.  Swagger is the toolset for helping architects write APIs quickly, and OAS 3.0 allows these APIs to be used be everyone.
```

Comment:
`AWS API Gateway will generate help you to test your API ` ?

Reply:
Ah yeah that's an awkward statement.  I meant to say that if you decide to use AWS API Gateway to manage/deploy your API, it's much easier if you already have a Swagger document.  

Will adjust.